pkg/products/monitoringspec/reconciler.go

```diff
@@ -0,0 +1,388 @@
+		}
+		serviceMonitors := &v1.ServiceMonitorList{}
+		err := serverClient.List(ctx, serviceMonitors, listOpts...)
+		if err != nil {
+			return integreatlyv1alpha1.PhaseFailed, err
+		}
+		for _, sm := range serviceMonitors.Items {
+			//Create a copy of service monitors in the monitoring namespace
+			//Create the corresponding rolebindings at each of the service namespace
+			key := sm.Namespace + `-` + sm.Name
+			delete(monSermonMap, key) // Servicemonitor exists, remove it from the local map
+			err := r.reconcileServiceMonitor(ctx, serverClient, sm)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, err
+			}
+		}
+	}
+
+	//Clean-up the stale service monitors and rolebindings if any
+	if len(monSermonMap) > 0 {
+		for _, sm := range monSermonMap {
+			//Remove servicemonitor
+			err = r.removeServicemonitor(ctx, serverClient, sm.Namespace, sm.Name)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, err
+			}
+			//Remove rolebindings
+			for _, namespace := range sm.Spec.NamespaceSelector.MatchNames {
+				err := r.removeRoleBinding(ctx, serverClient, namespace, roleBindingName)
+				if err != nil {
+					return integreatlyv1alpha1.PhaseFailed, err
+				}
+			}
+		}
+	}
+	return integreatlyv1alpha1.PhaseCompleted, err
+}
+
+func (r *Reconciler) reconcileServiceMonitor(ctx context.Context,
+	serverClient k8sclient.Client, serviceMonitor *v1.ServiceMonitor) (err error) {
+
+	if serviceMonitor.Spec.NamespaceSelector.Any {
+		logrus.Warnf("servicemonitor : %s cannot be copied to %s namespace. Namespace selector has been set to any",
+			serviceMonitor.Name, r.Config.GetNamespace())
+		return nil
+	}
+	sm := &v1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceMonitor.Namespace + `-` + serviceMonitor.Name,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+	opRes, err := controllerutil.CreateOrUpdate(ctx, serverClient, sm, func() error {
+		// Check if the servicemonitor has no  namespace selectors defined,
+		// if not add the namespace
+		sm.Spec = serviceMonitor.Spec
+		if len(sm.Spec.NamespaceSelector.MatchNames) == 0 {
+			sm.Spec.NamespaceSelector.MatchNames = []string{serviceMonitor.Namespace}
+		}
+		//Add all the original labels and append cloned servicemonitor label
+		sm.Labels = serviceMonitor.Labels
+		if len(sm.Labels) == 0 {
+			sm.Labels = make(map[string]string)
+		}
+		sm.Labels[clonedServiceMonitorLabelKey] = clonedServiceMonitorLabelValue
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Logger.Infof("operation result of creating servicemonitor %v was %v", sm.Name, opRes)
+
+	//Get the service monitor - that was created/updated
+	sermon := &v1.ServiceMonitor{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: sm.Name, Namespace: r.Config.GetNamespace()}, sermon)
+	if err != nil {
+		return err
+	}
+	//Create role binding for each of the namespace label selectors
+	for _, namespace := range sermon.Spec.NamespaceSelector.MatchNames {
```

Comment:
```suggestion
	//Create role binding for each of the namespace label selectors
	for _, namespace := range sm.Spec.NamespaceSelector.MatchNames {
```

I don't think it's needed to do a `serverClient.Get` again here -- everything you need should be already in the `sm` variable, shouldn't it?

Reply:
The mutate method which updates the servicemonitor object (the changes that is makes) is not available to the parent method that calls , createorupdae(). Never the less, the logic to create role bindings are split now, so it has get the servicemonitor and iterate over namespace selectors and reconcile the rolebindings for them.